Fix the to_storage for relative path

amritghimire opened the pull request:
This fixes the case where the path is passed to to_storage

skshetry commented:
Where does this relative path come from? Why don't we pass the full path to the `upload()`?

amritghimire commented:
> Where does this relative path come from? Why don't we pass the full path to the `upload()`?

https://iterativeai.slack.com/archives/C04A9RWEZBN/p1740628663390679

skshetry commented:
Not sure that I understand, could you please elaborate? I have read that thread.

amritghimire commented:
> Not sure that I understand, could you please elaborate? I have read that thread.

Updated the code. Basically the relative path is provided when running the script to to_storage. Moved the logic to to_storage instead.

skshetry commented:
This should be fixed in `export()` and/or `save()` as you could pass the relative paths there too, not in `to_storage()`.

Also, a few other things that I noticed:

1. Now that `export` supports exporting to any filesystems, it should not using `os.path`.


https://github.com/iterative/datachain/blob/88cee30f7ea495676a395b97bbcf69785f1be775/src/datachain/lib/file.py#L304


2. Should `export()` be renamed to be in symmetry with `to_storage()`?

skshetry commented on src/datachain/lib/file.py:
```diff
@@ -272,8 +272,12 @@ def read_text(self):
+        client: Client = self._catalog.get_client(destination)
+
+        if client.PREFIX == "file://" and not destination.startswith(client.PREFIX):
+            destination = f"{client.PREFIX}{destination}"
```
There is no such thing as relative `file://` URI. The path needs to be converted to an absolute path, and then to an URI.

skshetry commented on src/datachain/lib/file.py:
```diff
@@ -272,8 +272,12 @@ def read_text(self):
+        client: Client = self._catalog.get_client(destination)
+
+        if client.PREFIX == "file://" and not destination.startswith(client.PREFIX):
+            destination = client.PREFIX + os.path.abspath(destination)
```
This is still not correct. `file://` URI uses forward-slash as a separator, but by using `os.path`, the separator could be a backward-slash (`\\`) here.

You can use either of the options:
```python
from pathlib import Path
Path(destination).absolute().as_uri()
```

or, `urllib.request.pathname2url` or, `make_path_posix` from fsspec.

dreadatour reviewed (approved):
🙏 looks good to me!
